lib/src/components/NxFileUpload/NxMultiFileUpload/NxMultiFileUpload.tsx

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (c) 2019-present Sonatype, Inc.
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/.
+ */
+import React, { FormEvent, forwardRef, useRef, useContext, useEffect, useState } from 'react';
+import { faExclamationCircle, faTimesCircle } from '@fortawesome/free-solid-svg-icons';
+import classnames from 'classnames';
+import prettyBytes from 'pretty-bytes';
+import { forEach, last, without } from 'ramda';
+
+import NxFontAwesomeIcon from '../../NxFontAwesomeIcon/NxFontAwesomeIcon';
+import NxOverflowTooltip from '../../NxTooltip/NxOverflowTooltip';
+import NxButton from '../../NxButton/NxButton';
+import NxTooltip from '../../NxTooltip/NxTooltip';
+import { FormAriaContext } from '../../NxForm/context';
+import { useUniqueId } from '../../../util/idUtil';
+import { Props, propTypes, SelectedFileProps } from './types';
+
+export { Props };
+
+import '../NxFileUpload.scss';
+
+const formatSize = (size: number) => prettyBytes(size, { minimumFractionDigits: 1, maximumFractionDigits: 1 });
+
+function SelectedFile({ file, onDismiss: onDismissProp }: SelectedFileProps) {
```

Comment:
Isn't this just copied from `NxFileUpload`? They should share a single implementation of this part

Reply:
> Isn't this just copied from `NxFileUpload`? They should share a single implementation of this part

I wasn't sure how to proceed with this because in the multiple version, the `onDismiss` prop passes information back to NxMultiFileUpload's `onDismiss`,  there is a ref implemented in favour of the descriptionId on `nx-selected-file__info`, and now there's a useEffect for scrolling. Should it be a separate implementation?

Reply:
> I wasn't sure how to proceed with this because in the multiple version, the onDismiss prop passes information back to NxMultiFileUpload's onDismiss, there is a ref implemented in favour of the descriptionId on nx-selected-file__info, and now there's a useEffect for scrolling. Should it be a separate implementation?

`onDismiss` can probably be adjusted to function without passing the HTMLElement around. You're only using it to see if the dismissed file is the last one.  It should be possible to determine that using the `fileObj` and the `files`.

`descriptionId` can simply be optional.  I don't see `NxMultiFileUpload` setting `aria-describedby` at all (which may or may not be fine, we'll get to that later).

For the scrolling I'd recommend wrapping the shared `SelectedFile` component in a wrapper for `NxMultiFileUpload` that implements that `useEffect`.  `SelectedFile` will need to do ref forwarding for that to work.

Reply:
> `onDismiss` can probably be adjusted to function without passing the HTMLElement around. You're only using it to see if the dismissed file is the last one. It should be possible to determine that using the `fileObj` and the `files`.
> 
> `descriptionId` can simply be optional. I don't see `NxMultiFileUpload` setting `aria-describedby` at all (which may or may not be fine, we'll get to that later).
> 
> For the scrolling I'd recommend wrapping the shared `SelectedFile` component in a wrapper for `NxMultiFileUpload` that implements that `useEffect`. `SelectedFile` will need to do ref forwarding for that to work.

Implemented here: [2735b0d](https://github.com/sonatype/sonatype-react-shared-components/pull/850/commits/2735b0d32bdacdb312558cfa562a28ed7d2abcdd). I have left SelectedFile under NxFileUpload at this point, but can move it to its own file for both NxFileUpload and NxMultiFileUpload to import.

Reply:
> but can move it to its own file for both NxFileUpload and NxMultiFileUpload to import.

That's how I'd rather have it

Reply:
> > but can move it to its own file for both NxFileUpload and NxMultiFileUpload to import.
> 
> That's how I'd rather have it

great, moved here: [1f3f954](https://github.com/sonatype/sonatype-react-shared-components/pull/850/commits/1f3f954e18ff689c01d2fc3edfd1b0a0bba8342b)